bluepyopt/ephys/efeatures.py

```diff
@@ -226,3 +244,408 @@ def __str__(self):
+                np.asarray(self.exp_mean),
+                np.asarray(feature_value)
+            )
+
+        if np.isnan(score):
+            score = self.max_score
+
+        if self.force_max_score:
+            score = min(score, self.max_score)
+
+        logger.debug("Calculated score for %s: %f", self.name, score)
+
+        return score
+
+    def __str__(self):
+        """String representation"""
+
+        return ("%s for %s with stim start %s and end %s, "
+                "exp mean %s and std %s and AP threshold override %s"
+                % (self.extrafel_feature_name,
+                   self.recording_names,
+                   self.stim_start,
+                   self.stim_end,
+                   self.exp_mean,
+                   self.exp_std,
+                   self.threshold)
+                )
+
+
+def _interpolate_response(response, fs=20.0):
+    from scipy.interpolate import interp1d
+
+    x = response["time"]
+    y = response["voltage"]
+    f = interp1d(x, y, axis=1)
+    xnew = np.arange(np.min(x), np.max(x), 1.0 / fs)
+    ynew = f(xnew)  # use interpolation function returned by `interp1d`
+
+    response_new = {}
+    response_new["time"] = xnew
+    response_new["voltage"] = ynew
+
+    return response_new
+
+
+def _filter_response(response, fcut=[0.5, 6000], order=2, filt_type="lfilter"):
+    import scipy.signal as ss
+
+    fs = 1 / np.mean(np.diff(response["time"])) * 1000
+    fn = fs / 2.0
+
+    trace = response["voltage"]
+
+    if isinstance(fcut, (float, int, np.float, np.integer)):
+        btype = "highpass"
+        band = fcut / fn
+    else:
+        assert isinstance(fcut, (list, np.ndarray)) and len(fcut) == 2
+        btype = "bandpass"
+        band = np.array(fcut) / fn
+
+    b, a = ss.butter(order, band, btype=btype)
+
+    if len(trace.shape) == 2:
+        if filt_type == "filtfilt":
+            filtered = ss.filtfilt(b, a, trace, axis=1)
+        else:
+            filtered = ss.lfilter(b, a, trace, axis=1)
+    else:
+        if filt_type == "filtfilt":
+            filtered = ss.filtfilt(b, a, trace)
+        else:
+            filtered = ss.lfilter(b, a, trace)
+
+    response_new = {}
+    response_new["time"] = response["time"]
+    response_new["voltage"] = filtered
+
+    return response_new
+
+
+def _get_waveforms(response, peak_times, snippet_len_ms):
+    times = response["time"]
+    traces = response["voltage"]
+
+    assert np.std(np.diff(times)) < 0.001 * np.mean(
+        np.diff(times)
+    ), "Sampling frequency must be constant"
+
+    fs = 1.0 / np.mean(np.diff(times))  # kHz
+
+    reference_frames = (peak_times * fs).astype(int)
+
+    if isinstance(snippet_len_ms, (tuple, list, np.ndarray)):
+        snippet_len_before = int(snippet_len_ms[0] * fs)
+        snippet_len_after = int(snippet_len_ms[1] * fs)
+    else:
+        snippet_len_before = int((snippet_len_ms + 1) / 2 * fs)
+        snippet_len_after = int((snippet_len_ms - snippet_len_before) * fs)
```

Comment:
I find this implementation when `snippet_len_before` is a scalar strange, and it is not consistent with what is described in the docstring under `ms_cut`. Maybe use this:

`snippet_len_before = snippet_len_after = int(snippet_len_ms * fs)`

It would also be consistent with the case when `ms_cut` is an iterable.